File: strstr.go
// Time Complexity : O(m+n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
if needle is "" return 0
if needle is longer than haystack return -1
set i, j to 0
get lps array
loop over haystack with i
if haystack[i] == needle[j] increment i, j and if j==len(needle) return i-len(needle)
else if j > 0 && haystack[i] != needle[j] set j = lps[j-1]
else if j == 0 && haystack[i] != needle[j] increment i

return -1

longestCommonPrefixSuffix function
set j=0, i=1
loop over needle using i
if needle[i] == needle[j] increment j set lps[i]to j, increment i
else if j > 0 && haystack[i] != needle[j] set j = lps[j-1]
else if j == 0 && haystack[i] != needle[j] increment i
return lps
*/
package main

import "fmt"

func strStr(haystack string, needle string) int {
	if len(needle) == 0 {
		return 0
	}
	if len(haystack) < len(needle) {
		return -1
	}
	/*for i:=0;i<len(haystack) - len(needle) + 1;i++ {
	    if needle[0] == haystack[i] {
	        k := i
	        j := 0
	        for k < len(haystack) && needle[j] == haystack[k] {
	            k++
	            j++
	            if j == len(needle) {
	                return i
	            }
	        }
	    }
	}*/

	i := 0
	j := 0
	lps := longestCommonPrefixSuffix(needle)
	for i < len(haystack) {
		if haystack[i] == needle[j] {
			i++
			j++
			if j == len(needle) {
				return i - len(needle)
			}
		} else if j > 0 && haystack[i] != needle[j] {
			j = lps[j-1]
		} else if j == 0 && haystack[i] != needle[j] {
			i++
		}
	}


	return -1
}

func longestCommonPrefixSuffix(needle string) []int {
	j := 0
	i := 1
	lps := make([]int, len(needle))
	for i < len(needle) {
		//case 1
		if needle[i] == needle[j] {
			j++
			lps[i] = j
			i++
		} else if j > 0 && needle[i] != needle[j] {
			j = lps[j-1]
		} else if j == 0 && needle[i] != needle[j] {
			lps[i] = 0
			i++
		}
	}
	return lps
}

func MainStrStr() {
	fmt.Println(strStr("hello", "ll")) //expected 2
}

File: problem1/StrStr.java
// Time Complexity : O(m + n), m -> Length of haystack, n -> Length of needle
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem1;

public class StrStr {
	public int strStr(String haystack, String needle) {
		if (haystack == null || haystack.length() == 0 || needle == null || needle.length() == 0) {
			return -1;
		}
		int i = 0;
		int m = haystack.length();
		int n = needle.length();

		while (i < m) {
			int j = 0;

			// Store the value of i, so that it is not lost
			int ans = i;

			while (j < n && i < m && haystack.charAt(i) == needle.charAt(j)) {
				i++;
				j++;
			}

			// Needle found in Haystack
			if (j == n) {
				return ans;
			}

			// Reached end of haystack. Needle not found
			else if (i == m) {
				return -1;
			}

			// Not the end of either.
			else {
				i = ans + 1;
			}
		}
		return -1;
	}

	public static void main(String[] args) {
		StrStr obj = new StrStr();
		String haystack = "hello";
		String needle = "ll";
		System.out.println("Index of \'" + needle + "\' in \'" + haystack + "\' = " + obj.strStr(haystack, needle));
	}

}
